src/dump.c

```diff
@@ -290,7 +294,38 @@ static void jl_serialize_datatype(jl_serializer_state *s, jl_datatype_t *dt) JL_
 {
     int tag = 0;
     int internal = module_in_worklist(dt->name->module);
-    if (!internal && jl_unwrap_unionall(dt->name->wrapper) == (jl_value_t*)dt) {
+    if (mini_image && !jl_is_tuple_type(dt) && !jl_is_array_type(dt) && dt != jl_float64_type) {
+        if (dt->layout && jl_datatype_nfields(dt) > 0) {  // change the type to a tuple type with correct size
+            dt->name->module = jl_core_module;  // fudge the module
+            // jl_datatype_t *newdt = jl_apply_tuple_type(dt->types);
+            // newdt->layout = dt->layout;
+            // newdt->instance = dt->instance;
+            // newdt->ninitialized = dt->ninitialized;
+            // dt = newdt;
+        }
+        else if (dt->size == 1) {  // change the type to a UInt8
+            dt = jl_uint8_type;
+        }
+        else if (dt->size == 2) {  // change the type to a UInt16
+            dt = jl_uint16_type;
+        }
+        else if (dt->size == 4) {  // change the type to a UInt32
+            dt = jl_uint32_type;
+        }
+        else if (dt->size == 8) {  // change the type to a UInt64
+            dt = jl_uint64_type;
```

Comment:
I don't really understand this. Saving the wrong type tag for something doesn't seem useful?

Reply:
It was useful in the sense that it could make some code compile and run where it wouldn't compile before. The wrong type is often not a problem because the compiled code doesn't really use the type information, it just needs to get the size right (I know that this isn't always true, so it'd be nice not to do this). I was having problems where saving some types would cause the mini-image to explode in size as it tried to pull in more dependent modules and types. That is likely a sign of another problem, and maybe (hopefully) this is just a temporary band aid, but I haven't found the right approach.